pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -439,6 +446,28 @@ else let
         "/bin/sh"
       ];
       __propagatedImpureHostDeps = computedPropagatedImpureHostDeps ++ __propagatedImpureHostDeps;
+    } //
+    # If we use derivations directly here, they end up as build-time dependencies.
+    # This is especially problematic in the case of disallowed*, since the disallowed
+    # derivations will be built by nix as build-time dependencies, while those
+    # derivations might take a very long time to build, or might not even build
+    # successfully on the platform used.
+    # However, in order to scan for disallowed references in the final output,
+    # knowing the out path is sufficient, and the out path can be calculated from
+    # the derivation without needing to build it, saving time and resources.
+    # While the problem is less severe for allowed*, since we want the derivation
+    # to be built eventually, we would still like to get the error early and without
+    # having to wait while nix builds a derivation that might not be used.
+    # https://github.com/NixOS/nix/issues/4629
+    {
+      disallowedReferences =
+        map unsafeDerivationToUntrackedOutpath (attrs.disallowedReferences or [ ]);
+      disallowedRequisites =
+        map unsafeDerivationToUntrackedOutpath (attrs.disallowedRequisites or [ ]);
+      allowedReferences =
+        lib.mapNullable unsafeDerivationToUntrackedOutpath (attrs.allowedReferences or null);
+      allowedRequisites =
+        lib.mapNullable unsafeDerivationToUntrackedOutpath (attrs.allowedRequisites or null);
```

Comment:
Maybe (?) we could reduce the number of rebuilds by only adding these attributes if they were present if `attrs`

Reply:
Hmm yes this introduces the attrs to all derivations

```
nix-repl> bash.allowedRe
bash.allowedReferences
bash.allowedRequisites
nix-repl> bash.disallowedRe
bash.disallowedReferences
bash.disallowedRequisites
```

Also why do the allowed ref use or null

Reply:
They use `or null` since that it the no-op value for those attrs, an empty list would mean that nothing is allowed in the output.

Agreed for the other point though, we should not add the attributes if they were not present to begin with. I'll push a fix for that tomorrow.

Reply:
I pushed an update to fix this and to include the insights on the outPath in the comment.